### PR TITLE
Chair entity

### PR DESCRIPTION
**Pull request by chrisspieler**
Adds an ent_cinemachair entity, which may be tested on maps/ducc/cinema_main_ducc_01.vmap.

Currently, the chair supports sitting, lifting the armrests with the slot1 and slot2 buttons, and spawning cups with slot3 and slot4 when cinema.chair.debug is set to 1.

### CODE REVIEW

File: code/entities/chair/Armrest.cs
Comment (Remscar) [inline code/entities/chair/Armrest.cs:22]: Will an Armrest ever not have it's parent be the chair?
Could we make this a readonly property;
```
public CinemaChair Chair => Parent as CinemaChair;
```

File: code/entities/chair/Armrest.cs
Comment (Remscar) [inline code/entities/chair/Armrest.cs:26]: This might not happen, but I have concerns that the `State` variable could easily get out of sync.

It seems that `State` is really dependent on whether or not the armrest anim parameter `ArmrestToggleParameter` is true/false.

Is calling `GetAnimParameterBool` very expensive? We could just do
```
public States State => (States)Chair.GetAnimParameterBool(ArmrestToggleParameter)
```

Saves us the concern of tracking state, and simplifies the `HandleAnimTag` code.

File: code/entities/chair/Armrest.cs
Comment (Remscar) [inline code/entities/chair/Armrest.cs:65]: Would be more clear if this was `SetCupholdEntity` or maybe `SetCupholderEntity` or perhaps `Hold`/`HoldEntity` (was a bit confused what this was by it's name, I read it as "Cuphold an entity" but cuphold is not a verb)

Maybe even `PutEntityInCupholder` if you want to be verbose

File: code/entities/chair/Armrest.cs
Comment (Remscar) [inline code/entities/chair/Armrest.cs:71]: What happens if there is already an entity in the cupholder? I know right now you handle this with the user (caller) code, but we could handle that here too to be extra safe.

File: code/entities/chair/Armrest.cs
Comment (Remscar) [inline code/entities/chair/Armrest.cs:69]: Should this maybe return `false` to say "no we did not put the entity in the cupholder"?

File: code/entities/chair/Armrest.cs
Comment (Remscar) [inline code/entities/chair/Armrest.cs:77]: Nit: Move properties/members to the top of the class before any methods

File: code/entities/chair/Armrest.cs
Comment (Remscar) [inline code/entities/chair/Armrest.cs:80]: Nit: Consider naming `LaunchContents` or `LuanchCupholderContents` or whatever you end up renaming `Cupholded` to.

File: code/entities/chair/CinemaChair.cs
Comment (Remscar) [inline code/entities/chair/CinemaChair.cs:66]: Consider just doing
```
public string CannotUseText => IsOccupied ? "Seat is occupied" : null;
```

File: code/entities/chair/CinemaChair.cs
Comment (Remscar) [inline code/entities/chair/CinemaChair.cs:31]: These could cause errors on the client while the Dictionary is propagating through the network.

If you try to get a value from a dictionary with `[]` and the key doesn't exist, an exception is thrown.
Instead you can do this:
```
public Armrest LeftArmrest
        => Armrests.TryGetValue(Armrest.Sides.Left, out Armrest value) ? value : null;
```
And null will be returned instead.

File: code/entities/chair/CinemaChair.cs
Comment (Remscar) [inline code/entities/chair/CinemaChair.cs:114]: Please remove log statements

File: code/entities/chair/CinemaChair.cs
Comment (Remscar) [inline code/entities/chair/CinemaChair.cs:125]: Please remove this log statement

File: code/entities/chair/CinemaChair.cs
Comment (Remscar) [inline code/entities/chair/CinemaChair.cs:129]: Instead of throwing an exception could we just return?

File: code/entities/chair/CinemaChair.cs
Comment (Remscar) [inline code/entities/chair/CinemaChair.cs:120]: Performance best practice would be to check this first and return false if it's occupied.

Then, check the angle. Reason being is that `.Normal` requires a square root operation which is expensive.

File: code/entities/chair/CinemaChair.cs
Comment (Remscar) [inline code/entities/chair/CinemaChair.cs:185]: Nit: Scope pyramids could be reduced with:
```
var armrest = tag.Contains("left") ? LeftArmrest : RightArmrest;
armrest.HandleAnimTag(tag);
```

File: code/player/controllers/ChairController.cs
Comment (Remscar) [inline code/player/controllers/ChairController.cs:45]: Please remove debug log statements.

File: code/player/controllers/ChairController.cs
Comment (Remscar) [inline code/player/controllers/ChairController.cs:59]: Nit: could remove the else by making this `return` on completion.

File: code/player/controllers/ChairController.cs
Comment (Remscar) [inline code/player/controllers/ChairController.cs:71]: I'm assuming this will be removed later so i'm ignoring any debug slot / test code for this.

**Review by Remscar — CHANGES_REQUESTED** [review]
This is looking great! A lot of really good code in this PR.

Ejecting the drinks was a lot of fun when I tested it, though there was one case where once I got out of the seat, I couldn't get back in. I can't see an obvious bug in the code so we'll have to see if it ever happens again.

Besides some code hygiene changes and safety checks I think this is good to go!

**Review by Remscar — CHANGES_REQUESTED** [review]
This is looking great! A lot of really good code in this PR.

Ejecting the drinks was a lot of fun when I tested it, though there was one case where once I got out of the seat, I couldn't get back in. I can't see an obvious bug in the code so we'll have to see if it ever happens again.

Besides some code hygiene changes and safety checks I think this is good to go!

File: code/entities/chair/Armrest.cs
Comment (chrisspieler) [reply code/entities/chair/Armrest.cs]: Made a change so that an already held entity will be launched. This uses a different launch direction and impulse magnitude, so I made LaunchHeldEntity use parameters for this.

https://github.com/tech-nawar/sbox-cinema/assets/8889285/1b098263-b028-435c-a8ab-aa78c03c53a1

File: code/entities/chair/Armrest.cs
Comment (chrisspieler) [reply code/entities/chair/Armrest.cs]: I went ahead and changed this method to `TryHoldEntity`, which returns whether the `entity` argument actually gets held.

File: code/entities/chair/CinemaChair.cs
Comment (chrisspieler) [reply code/entities/chair/CinemaChair.cs:36]: Implemented the proposed change and added various null checks for callers of Armrests, LeftArmrest, and RightArmrest.

File: code/entities/chair/Armrest.cs
Comment (chrisspieler) [reply code/entities/chair/Armrest.cs]: I removed the intermediate states of Armrest.States, so now the only remaining states are Raised and Lowered, which correspond roughly to the aforementioned anim parameter. Now it's possible to set `HeldEntity` while the arm is still being lowered, which is probably fine.

File: code/entities/chair/Armrest.cs
Comment (Remscar) [inline code/entities/chair/Armrest.cs:79]: This could lead to a potential race condition if the `LaunchHeldEntity` doesn't execute immediately, or the default parameter `disableCollisionTime` is ever changed from 0. I'm not sure if the former would ever happen in C#, so maybe we can ignore that.

Either way, this isn't blocking, just a warning of a potential bug in the future if the `LaunchHeldEntity` code changes very much.

**Review by Remscar — APPROVED** [review]
Looks great, this has my 👍 

Feel free to merge when you like or wait for another review from Rifter or Walker

**Pull request merged**
